### PR TITLE
Fixed revert failure due to OS file locks by Git

### PR DESCRIPTION
**Pull request by rweber89**
Git reset acquires a file lock that's not release in time in half the cases before git checkout runs, making git checkout fail.
I added a loop of tries and a sleep to make sure it succeeds. From testing this a handful of times, on my machine it succeeds either at first or second try.

### CODE REVIEW

File: Source/GitSourceControl/Private/GitSourceControlUtils.cpp [hunk 1]
```diff
@@ -2105,9 +2105,7 @@ bool PullOrigin(const FString& InPathToGitBinary, const FString& InPathToReposit
 
 	// Get the list of files which will be updated (either ones we changed locally, which will get potentially rebased or merged, or the remote ones that will update)
 	TArray<FString> DifferentFiles;
-	TArray<FString> Parameters {TEXT("--name-only"), RemoteBranch};
-	const bool bResultDiff = RunCommand(TEXT("diff"), InPathToGitBinary, InPathToRepositoryRoot, Parameters, FGitSourceControlModule::GetEmptyStringArray(),
-										DifferentFiles, OutErrorMessages);
+	const bool bResultDiff = RunCommand(TEXT("diff"), InPathToGitBinary, InPathToRepositoryRoot, { TEXT("--name-only"), RemoteBranch }, FGitSourceControlModule::GetEmptyStringArray(), DifferentFiles, OutErrorMessages);
```
Comment (mastercoms) [inline Source/GitSourceControl/Private/GitSourceControlUtils.cpp:2108]: Could you revert this change? I'm pretty sure this is a Microsoft extension.

File: Source/GitSourceControl/Private/GitSourceControlOperations.cpp [hunk 2]
```diff
@@ -525,7 +528,20 @@ bool FGitRevertWorker::Execute(FGitSourceControlCommand& InCommand)
 		if (OtherThanAddedExistingFiles.Num() > 0)
 		{
 			// revert any changes in working copy (this would fails if the asset was in "Added" state, since after "reset" it is now "untracked")
-			InCommand.bCommandSuccessful &= GitSourceControlUtils::RunCommand(TEXT("checkout"), InCommand.PathToGitBinary, InCommand.PathToRepositoryRoot, FGitSourceControlModule::GetEmptyStringArray(), OtherThanAddedExistingFiles, InCommand.ResultInfo.InfoMessages, InCommand.ResultInfo.ErrorMessages);
+			// may need to try a few times due to file locks from prior operations
+			bool CheckoutSuccess = false;
+			for (int32 i = 0; i < 10; ++i)
```
Comment (mastercoms) [inline Source/GitSourceControl/Private/GitSourceControlOperations.cpp]: I'd prefer

```cpp
int32 Tries = 10;
while (Tries-- > 0)
```

File: Source/GitSourceControl/Private/GitSourceControlOperations.cpp [hunk 3]
```diff
@@ -525,7 +528,20 @@ bool FGitRevertWorker::Execute(FGitSourceControlCommand& InCommand)
...
+			{
+				CheckoutSuccess = GitSourceControlUtils::RunCommand(TEXT("checkout"), InCommand.PathToGitBinary, InCommand.PathToRepositoryRoot, FGitSourceControlModule::GetEmptyStringArray(), OtherThanAddedExistingFiles, InCommand.ResultInfo.InfoMessages, InCommand.ResultInfo.ErrorMessages);
+				if (CheckoutSuccess)
+				{
+					break;
+				}
+
+				std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (mastercoms) [inline Source/GitSourceControl/Private/GitSourceControlOperations.cpp]: Use `FPlatformProcess::Sleep(0.1f);` (include: `#include "HAL/PlatformProcess.h"`)

**Review by mastercoms — CHANGES_REQUESTED** [review]

File: Source/GitSourceControl/Private/GitSourceControlOperations.cpp (hunk 3 above)
Comment (rweber89) [reply Source/GitSourceControl/Private/GitSourceControlOperations.cpp]: Done

File: Source/GitSourceControl/Private/GitSourceControlOperations.cpp (hunk 2 above)
Comment (rweber89) [reply Source/GitSourceControl/Private/GitSourceControlOperations.cpp]: Done

File: Source/GitSourceControl/Private/GitSourceControlUtils.cpp (hunk 1 above)
Comment (rweber89) [reply Source/GitSourceControl/Private/GitSourceControlUtils.cpp:2108]: It's a language feature - I think since C++11

Comment (ameaninglessname) [reply Source/GitSourceControl/Private/GitSourceControlUtils.cpp:2108]: yeah, as for as i known, {} introduced in c++11 used to unified the initialization expression. 

One small thing may be this line is pretty long and need two lines for it.

Comment (mastercoms) [reply Source/GitSourceControl/Private/GitSourceControlUtils.cpp:2108]: I meant the reference. https://docs.microsoft.com/en-us/cpp/build/reference/zc-referencebinding-enforce-reference-binding-rules?view=msvc-170

Comment (rweber89) [reply Source/GitSourceControl/Private/GitSourceControlUtils.cpp:2108]: Yup that's what I meant. The link is only for binding to references. The function argument is a const ref, which extends the lifetime of the temporary to the scope that it is referenced in, so here the body of the function.

File: Source/GitSourceControl/Private/GitSourceControlOperations.cpp [hunk 4]
```diff
@@ -530,15 +531,16 @@ bool FGitRevertWorker::Execute(FGitSourceControlCommand& InCommand)
 			// revert any changes in working copy (this would fails if the asset was in "Added" state, since after "reset" it is now "untracked")
 			// may need to try a few times due to file locks from prior operations
 			bool CheckoutSuccess = false;
-			for (int32 i = 0; i < 10; ++i)
+			int32 Attempts = 0;
+			while( --Attempts >= 0 )
```
Comment (mastercoms) [inline Source/GitSourceControl/Private/GitSourceControlOperations.cpp]: Attempts is init to 0 instead of 10, and I think it's better to go with post increment and > than pre and >=

**Review by mastercoms — CHANGES_REQUESTED** [review]

File: Source/GitSourceControl/Private/GitSourceControlOperations.cpp (hunk 4 above)
Comment (rweber89) [reply Source/GitSourceControl/Private/GitSourceControlOperations.cpp]: Done

**Review by mastercoms — APPROVED** [review]
